include/inviwo/core/io/datareaderfactory.h

```diff
@@ -42,44 +42,62 @@ namespace inviwo {
     virtual ~DataReaderFactory() = default;
 
     bool registerObject(DataReader* reader);
     bool unRegisterObject(DataReader* reader);
     virtual std::unique_ptr<DataReader> create(const FileExtension& key) const override;
-    virtual std::unique_ptr<DataReader> create(const std::string& key) const;
-    virtual bool hasKey(const std::string& key) const;
+    virtual std::unique_ptr<DataReader> create(std::string_view key) const;
+    virtual bool hasKey(std::string_view key) const;
     virtual bool hasKey(const FileExtension& key) const override;
 
     template <typename T>
     std::vector<FileExtension> getExtensionsForType() const;
 
+    /**
+     * \brief Return a reader matching the file extension of DataReader of type T.
+     * Does case insensive comparison between the last part of filePathOrExtension and each
+     * registered extension.
+     * @param filePathOrExtension Path to file, or simply the extension.
+     * @return First available DataReaderType<T> if found, nullptr otherwise.
+     */
     template <typename T>
-    std::unique_ptr<DataReaderType<T>> getReaderForTypeAndExtension(const std::string& ext) const;
+    std::unique_ptr<DataReaderType<T>> getReaderForTypeAndExtension(
+        std::string_view filePathOrExtension) const;
     template <typename T>
     std::unique_ptr<DataReaderType<T>> getReaderForTypeAndExtension(const FileExtension& ext) const;
 
     /**
      * First look for a reader using the FileExtension ext, and if no reader was found look for a
-     * reader using the fallbackExt. This is often used with a file open dialog, where the dialog
-     * will have a selectedFileExtension that will be used as ext, and the fallbackExt is taken from
-     * the file to be opened. This way any selected reader will have priority over the file
-     * extension.
+     * reader using the fallbackFilePathOrExtension. This is often used with a file open dialog,
+     * where the dialog will have a selectedFileExtension that will be used as ext, and the
+     * fallbackFilePathOrExtension is taken from the file to be opened. This way any selected reader
+     * will have priority over the file extension.
      */
     template <typename T>
     std::unique_ptr<DataReaderType<T>> getReaderForTypeAndExtension(
-        const FileExtension& ext, const std::string& fallbackExt) const;
+        const FileExtension& ext, std::string_view fallbackFilePathOrExtension) const;
 
     template <typename T>
-    bool hasReaderForTypeAndExtension(const std::string& ext) const;
+    bool hasReaderForTypeAndExtension(std::string_view filePathOrExtension) const;
     template <typename T>
     bool hasReaderForTypeAndExtension(const FileExtension& ext) const;
 
 protected:
-    Map map_;
+    // Sort extensions by length to first compare long extensions. Avoids selecting extension xyz in
+    // case xyzw exist, see getReaderForTypeAndExtension
+    Map map_ = Map([](const auto& a, const auto& b) -> bool {
```

Comment:
> Will this not change the order of all the readers in the dropdown in Inviwo?

No, it was an unordered map before. 

> Also I find this solution a bit weird, I think it would be much cleaner to handle all this at the match stage.

Agree, it would be cleaner but it would also be unnecessarily expensive. Would need to go through all readers during retrieval or sort them each time. Not sure it is worth the extra implementation time

Reply:
you only have to select the longest matching one, and there should be very few of those..  but fine as long as the order is not mixed up. Do we sort the list in the GUI instead then?

Reply:
Yes, the list is sorted in the GUI. I checked that because I had the same thought.